language/move-lang/src/diagnostics/codes.rs

```diff
@@ -172,6 +172,11 @@ codes!(
             { msg: "cyclic type instantiation", severity: NonblockingError },
         MissingAcquires: { msg: "missing acquires annotation", severity: NonblockingError },
         InvalidNum: { msg: "invalid number after type inference", severity: NonblockingError },
+        NonInvocablePublicScript: {
+            msg: "script function cannot be invoked with this signature \
```

Comment:
Riffing on the error message with some common gotchas:

```
Invalid script function. A valid script function must:
- begin with one or more `signer`'s
- contain no references (e.g., `&signer`)
- contain no user-defined struct types (e.g., `Coin`)
- have no return type
```

This isn't an exhaustive description of the rules, but I think it will cover the most common mistakes.

Reply:
You want all of this in the error message block?

Reply:
Yeah--I feel like if we toss out this error without explaining how to fix it, we haven't fully fixed the UX footgun. 

It would be less verbose (and more cool) if we tell the user which rule(s) their `script` function is violating, but that's obviously more work. I don't know enough about the compiler's error reporting structure to know whether that's feasible.

Reply:
You do get each one of these explained per type usecase FWIW.

```
warning[W04022]: script function cannot be invoked with this signature (NOTE: this may become an error at the bytecode level in the future)
  ┌─ tests/move_check/typing/public_script_signature_warning.move:7:24
  │
7 │     public(script) fun signer_ref(_: &signer) {}
  │                        ^^^^^^^^^^    ------- Found: '&signer'. But expected one of: 'u8', 'u64', 'u128', 'bool', 'address', 'vector<_>'
  │                        │              
  │                        Invalid parameter for script function 'signer_ref'

warning[W04022]: script function cannot be invoked with this signature (NOTE: this may become an error at the bytecode level in the future)
  ┌─ tests/move_check/typing/public_script_signature_warning.move:9:24
  │
9 │     public(script) fun late_signer(_u: u64, _s: signer) {}
  │                        ^^^^^^^^^^^              ------ 'signer's must be a prefix of the arguments to a script--they must come before any non-signer types
  │                        │                         
  │                        Invalid parameter for script function 'late_signer'

warning[W04022]: script function cannot be invoked with this signature (NOTE: this may become an error at the bytecode level in the future)
   ┌─ tests/move_check/typing/public_script_signature_warning.move:11:24
   │
11 │     public(script) fun struct_arg(_: CoolStruct) {}
   │                        ^^^^^^^^^^    ---------- Found: '0x42::M::CoolStruct'. But expected one of: 'u8', 'u64', 'u128', 'bool', 'address', 'vector<_>'
   │                        │              
   │                        Invalid parameter for script function 'struct_arg'

warning[W04022]: script function cannot be invoked with this signature (NOTE: this may become an error at the bytecode level in the future)
   ┌─ tests/move_check/typing/public_script_signature_warning.move:13:24
   │
13 │     public(script) fun u64_ret(): u64 {
   │                        ^^^^^^^    --- Given: 'u64'
   │                        │           
   │                        'public(script)' functions must have a return type of '()' in order to be invocable as a script entry point
   │                        Expected: '()'

warning[W04022]: script function cannot be invoked with this signature (NOTE: this may become an error at the bytecode level in the future)
   ┌─ tests/move_check/typing/public_script_signature_warning.move:17:24
   │
17 │     public(script) fun struct_ret(): CoolStruct {
   │                        ^^^^^^^^^^    ---------- Given: '0x42::M::CoolStruct'
   │                        │              
   │                        'public(script)' functions must have a return type of '()' in order to be invocable as a script entry point
   │                        Expected: '()'
```

Reply:
Ah, nice. That definitely helps!

Reply:
Ah nice. That definitely helps!

Given that, I'd suggest leaving the message as-is except for dropping the part about the bytecode. Doesn't seem like the user needs to know about the bytecode--we could say something like "This may become an error in the future", or say nothing + keep it short and sweet.

Reply:
> "This may become an error in the future"

This is what I had before, and so I'll go back to that. 

I want to get in the habit of marking these sorts of potentially breaking changes early in the source language, so this phrasing works